### PR TITLE
Feature/improve phrase

### LINKED ISSUE #1512: Grammatically improve sentences from neural boards using AI
New Cboard feature using IA

Grammatically improve sentences from neural boards using AI

The phrases armed through neural boards, aim to communicate in a simple way. These sentences can be grammatically improved using artificial intelligence.

Looking at this use case:

image3

An example using GPT is shown

prompt:

3 options to grammatically improve this sentence: “what you to think”

Response:

    What do you think?
    What is it that you are thinking?
    What are your thoughts on this matter?

Another possible feature is adding the possibility to change the time of the phrase from a button of the UI

prompt:

        rewrite it in the past tense.

Response:

"What did you think?"

prompt::

        rewrite it in the future tense.

response:

"What will you think?"

Current development

Currently added a space in the output bar to display the sentence that was improved. Also a "play" button to say this phrase. If you want to reproduce the phrase raw, you can do so by clicking on the output bar. As our app does natively.

image2

Disadvantages

    Only one option can be accessed for the phrase.
    Present, past, and future tenses are not selectable

Advantages

    The User Interface is simple
    It is possible to reproduce the phrase in raw

Possible variations

Replacing the “play” button for the phrase. We can include a button to replace it with the next option, and thus take a tour of the different options, including the raw option. The phrase will be played by clicking on the output bar.

image1

Disadvantages

    Using the feature is more complex
    Playing the raw phrase requires more clicks

Advantages

    It is possible to access more options to improve the phrase

### PR DESCRIPTION
**Pull request by tomivm**
Should use this branch for the API https://github.com/cboard-org/cboard-api/tree/feature/GPT-edit-phrases
Close #1512

### CODE REVIEW

File: src/components/Board/__tests__/Board.actions.test.js [hunk 1]
```diff
@@ -200,11 +203,18 @@ describe('actions', () => {
...
+      {
+        type: types.CHANGE_OUTPUT,
+        output
+      }
+    ];
+
+    const store = mockStore(initialState);
+    store.dispatch(actions.changeOutput(output));
+    const dispatchedActions = store.getActions();
+    console.log(dispatchedActions);
```
Comment (RodriSanchez1) [inline src/components/Board/__tests__/Board.actions.test.js]: remove log

File: src/providers/SubscriptionProvider/SubscriptionProvider.actions.js [hunk 2]
```diff
@@ -43,6 +44,8 @@ export function updateIsOnTrialPeriod() {
         isOnTrialPeriod
       })
     );
+
+    if (!isOnTrialPeriod) disablePremiumFeaturesOnTrialPeriodEnded();
```
Comment (RodriSanchez1) [inline src/providers/SubscriptionProvider/SubscriptionProvider.actions.js]: I think this will also disable the feature in users who are premium

Comment (tomivm) [reply src/providers/SubscriptionProvider/SubscriptionProvider.actions.js]: Really on truth. Good Review!

**Review by martinbedouret — CHANGES_REQUESTED** [review]
@tomivm We are going to update the UI for this feature in order to get a cleaner output bar. The idea is to add a suggestion bar below the board as the following image: 

image

File: src/components/Board/ImprovePhraseOutput/ImprovePhraseOutput.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import styles from './ImprovePhraseOutput.module.css';
+
+import { Typography } from '@material-ui/core';
+import { PlayArrow } from '@material-ui/icons';
+
+const propTypes = {
+  isRootBoardTourEnabled: PropTypes.bool,
+  isUnlockedTourEnabled: PropTypes.bool,
+  isLocked: PropTypes.bool,
+  disableTour: PropTypes.func.isRequired
```
Comment (RodriSanchez1) [inline src/components/Board/ImprovePhraseOutput/ImprovePhraseOutput.js]: `Warning: Failed prop type: The prop disableTour is marked as required in ImprovePhraseOutput, but its value is undefined.`

**Review by RodriSanchez1 — CHANGES_REQUESTED** [review]
We need to handle the ui when text is large. See: 
image


When clearing all tiles with the backspace button, the improved phrase shows the last request instead of being empty.
image

File: src/components/Board/Board.actions.js [hunk 4]
```diff
@@ -324,9 +325,40 @@ export function clickOutput(outputPhrase) {
...
+  return async (dispatch, getState) => {
+    try {
+      const language = getState().language.lang;
+      const improvedPhrase = await improvePhrase(language);
+      dispatch({
+        type: CHANGE_IMPROVED_PHRASE,
+        improvedPhrase
+      });
+    } catch (err) {
+      console.log('error', err);
```
Comment (RodriSanchez1) [inline src/components/Board/Board.actions.js]: Change for `console.error('Error: ', err.message);`

**Comment by tomivm** [comment]
Thanks, @RodriSanchez1 I added a new capability to cancel every old request before a new request to improve the phrase starts. This solves the problem 'When clearing all tiles with the backspace button, the improved phrase shows the last request instead of being empty.' and improves the feature's performance. See [2d033a5](https://github.com/cboard-org/cboard/pull/1513/commits/2d033a5e014cae1066b00762743586ae5286828a)

File: src/components/Settings/Navigation/Navigation.component.js [hunk 5]
```diff
@@ -80,6 +81,12 @@ class Navigation extends React.Component {
     });
   };
 
+  toggleimprovePhraseActive = () => {
```
Comment (shayc) [inline src/components/Settings/Navigation/Navigation.component.js]: Nit: should be `toggleImprove...` capital I

File: src/components/Board/Board.actions.js [hunk 6]
```diff
@@ -324,9 +326,42 @@ export function clickOutput(outputPhrase) {
...
+
+export function improvePhrase(output) {
+  const improvePhrase = async language => {
+    const MIN_TILES_TO_IMPROVE = 1;
+    if (output.length <= MIN_TILES_TO_IMPROVE) return '';
+    const labels = output.map(symbol => symbol.label);
+    const phrase = labels.join(' '); //this.handlePhraseToShare();
+    const improvedPhrase = await API.improvePhrase({ phrase, language });
+    return improvedPhrase.phrase;
+  };
```
Comment (shayc) [inline src/components/Board/Board.actions.js]: `improvePhrase` is being used for both export fn and the nested const and also the slightly modified `improvedPhrase` in the return. 

Nothing wrong from code perspective, it's just harder to read..

File: src/components/Board/ImprovePhraseOutput/ImprovePhraseOutput.module.css [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  border: 2px solid transparent;
+  height: 3.2em;
+  background-color: indigo;
+  color: white;
+}
+
+.text_and_controls.enabled:hover,
+.text_and_controls.enabled:focus {
+  outline: none;
+  border: 2px solid #ffffff; /* Red border */
```
Comment (shayc) [inline src/components/Board/ImprovePhraseOutput/ImprovePhraseOutput.module.css]: The comment is a lie 😀

File: src/components/Board/Output/SymbolOutput/ImprovePhraseControls.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    <div style={{ position: 'relative' }}>
+      <div
+        className={`SymbolOutput__improved__Phrase__controls ${bigControlsClassName}`}
+      >
+        <div className="improved__Phrase_shadow" />
+        <div className="improved_Phrase_Button">
+          <IconButton
+            color="inherit"
+            className={
+              increaseOutputButtons
```
Comment (shayc) [inline src/components/Board/Output/SymbolOutput/ImprovePhraseControls.js]: Name suggests this is a function (verb) but used as Boolean. For readability Boolean values usually starts with is/should `shouldIncreaseOutputButtons`

**Review by shayc — APPROVED** [review]
Looks good, some minor feedback

**Comment by tomivm** [comment]
Thanks @shayc 🚀

**Pull request merged**